### ISSUE TITLE
Implement cacheing

### ISSUE THREAD

**Issue by nadrimajstor** [adelton/mod_authnz_pam#1]
Please consider implementing some sort of cacheing as using AuthBasicProvider against FreeIPA makes subversion quite slow to work with.

**Comment by adelton** [comment]
Thanks for the suggestion. I'll try to look into adding support for mod_authn_socache per https://httpd.apache.org/docs/2.4/mod/mod_authn_socache.html#dev.

**Comment by schnitzi99** [comment]
Thanks for the module and your work on that!
I added support for caching via socache and it seems to work so far.

I don't have any programming experience (except playing around some 20 years ago) and was pretty much muddling through looking at other code examples. So please thoroughly check and correct my adaptions before using / including them. File including changes is attached to this message.

How does it work:
Socache seems to be comparing a cached version (needs to be hashed) of the password with the 
typed-in-password. Therefore we need to store the a salted md5 hash of the password in the socache when authorization was successful. In my implementation the salt is predefined - I guess a random salt would be better.

What configuration is necessary:
Apache-config needs at least the following (upper case is necessary) for it to work:
AuthBasicProvider socache PAM
AuthnCacheProvideFor PAM
--> With this access on a complex page causes only one access via PAM and after that socache takes over as long as the cache is not expired.

**Comment by schnitzi99** [comment]
I cant seem to attach the file to this post. Can I somehow send an e-mail?

**Comment by adelton** [comment]
@schnitzi99: Please open a pull request -- that way we will be able to talk about the change in public. Thank you.

**Comment by adelton** [comment]
I've created one at https://github.com/adelton/mod_authnz_pam/pull/2 now, to start discussion. I did not test it, I did not even try to compile it -- just applied the changes from trunk to mod_authnz_pam code.

**Comment by adelton** [comment]
Support for

```
	AuthBasicProvider socache PAM
	AuthnCacheProvideFor PAM 
```

on Apache 2.4 is now in master. Thanks to @schnitzi99 for the initial code.

**Comment by KrzysztofHajdamowicz** [comment]
Hello!

I'm sorry to reopen this issue, but it conflicts with usage of one time passwords used by pam-yubico.
We are using pam-yubico to perform 2FA on Zabbix Frontend.

Problem is, entering one-time-password is checked against stored database and next request is validated again, against same one-time password, thus giving "password replayed" error.
I've tried with `pam_timestamp`, but no luck. Then, I've found this issue.

Caching password only wouldn't work in this usecase as whole authentication should be cached for several minutes.
